database/src/mdbx/metrics_handler.rs

```diff
@@ -0,0 +1,132 @@
+// This file has been ported and adapted from reth (https://github.com/paradigmxyz/reth).
+
+use std::{
+    backtrace::Backtrace,
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use libmdbx::{CommitLatency, TransactionKind};
+use log::{trace, warn};
+
+use crate::metrics::{DatabaseEnvMetrics, TransactionMode, TransactionOutcome};
+
+/// Du
+const LONG_TRANSACTION_DURATION: Duration = Duration::from_secs(60);
+
+#[derive(Debug)]
+pub struct MetricsHandler<K: TransactionKind> {
+    /// Some identifier of transaction.
+    txn_id: u64,
+    /// The time when transaction was opened.
+    start: Instant,
+    /// If `true`, the metric about transaction closing has already been recorded and we don't need
+    /// to do anything on [`Drop::drop`].
+    close_recorded: bool,
+    /// If `true`, the backtrace of transaction will be recorded and logged.
+    /// See [`MetricsHandler::log_backtrace_on_long_read_transaction`].
+    record_backtrace: bool,
+    /// If `true`, the backtrace of transaction has already been recorded and logged.
+    /// See [`MetricsHandler::log_backtrace_on_long_read_transaction`].
+    backtrace_recorded: AtomicBool,
+    pub(super) env_metrics: Arc<DatabaseEnvMetrics>,
+    _tx_kind: PhantomData<K>,
```

Comment:
```suggestion
    transaction_mode: TransactionMode,
```
would probably be better, then there aren't two instantiations of this struct.